### DISCUSSION TITLE
Interface implementation syntax

### DISCUSSION THREAD

**Discussion by cmxog** [tact-lang/spec#6] [Polls]
Current proposals:

* [Separate impl](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2717067)
* [Consolidated method surface](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2717092)

Resolved as in https://github.com/tact-lang/docs/discussions/6#discussioncomment-3062359.

**Comment by oleganza** [comment]
>The above syntax enables method implementation to be spread out throughout the program and fails to deliver a comprehensive view of what's available.

Is this the only critique of the Rust-style impls? IMHO it's easy to enumerate all impls together nicely, except Rust's way (vs Swift/Java way) is a bit extra declarative boilerplate across groups of methods.

Separate impls allow consistent way to impl interfaces on other types. Also, it's always easy to combine together separate things than to untangle entities afterwards. Hence, my superficial preference for separate impls — unless there are some concrete pitfalls with them.

**Comment by cmxog** [comment]
# Separate `impl`s

One way to unify method implementation syntax is to follow Rust-inspired model of `impl`s, both for implementing methods on a type as well as implementing interfaces

```
impl MyType {
   fn ...
}
```

```
impl MyInterface for MyType {
}
```

#### Cons:

* Unclear how to implement parameterization (see @p0lunin's comment)
* Scattered implementation surface
* Potential unfamiliarity

#### Pros:

* Method grouping by interface (or lack thereof)
* Ability to implement interface on existing types (including primitive types such as `Int257`)
* Removes the dependence on the `{}` syntax for some types like [union](https://github.com/tact-lang/spec/discussions/4#discussioncomment-2716819)

**Reply by p0lunin** [reply]
There also exists one significant problem with such approach. In Rust impls with generics have syntax `impl<Generics> for ...` which is ok for contraint-based language such as rust because it translates to the rule `forall Generics. impl ...`. In Tact generics is compile-time functions that generates output types. So, we cannot just reuse Rust syntax without resolving this problem. Maybe we can extend existing types with such impl or we can provide additional semantic that allow `impl`s on generic types?

**Reply by cmxog** [reply]
That is a very good catch indeed. I guess this is what consolidated syntax may capture better as method/function definitions are still within the same parameterization scope and what gets produces is the type with methods specialized for this.

**Comment by cmxog** [comment]
# Consolidated method surface

What if we consolidated the entirety of methods in `struct` (extending what we have right now):

```
struct {
  fn method() ...
} // this one only has its own methods

struct : Interface1, Interface2 {
   ...
} // has to implement methods from Interface1 and Interface2

struct (X: Type) : Interface {
} // same syntax applies to shorthands

// An wordier but clearer syntax:
struct implements Interface, Interface2 {

}

// with a shorthand
struct impl Interface, Interface2 {

}

```

This way we can always tell exactly what the struct implements. 

The main drawback that I can see is that it might not be clear *which exact* methods in the implementation are coming from which interface as they are all lumped together. There are a few ways to mitigate this:

* Error messages about missing methods should indicate which interface are they declared in
*  Consider namespacing methods of interfaces like:

```
struct : Iface {
  fn Iface::method() ...
}
```

#### Cons:

* Noisy method surfaces
* Tangling data and code together in one piece
* Impossible to use interfaces on existing data types

#### Pros:

* Potentially, much simpler parameterization story as everything is included within the comptime function defining both the type and all impls (see @p0lunin's [comment](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2717909))
* No scattering of implementation details

**Reply by cmxog** [reply]
We can potentially improve method grouping characteristics in this approach if we didn't declare all implemented interface in the head of the construct (`struct : Iface, ... { ... }`) but instead left it in-bounds of the details. 

Here's an example:

```
struct {
  fn own_method(..) { .. }

  impl Iface {
     fn iface_method(..) { .. }
  }

}
```

This way methods are still grouped but not scattered.

**Reply by cmxog** [reply]
For implementations of interface on "externally-defined" (or builtin) types, we may want to consider stipulating newtype semantics. This way, one won't be able to implement interfaces on types they don't "own" but if we will have shortcuts for simple newtypes, it may be a passable solution.

Let's say we wrap `Int257` with `MyInt257` and then we can implement anything we want on it. As long as we have relatively easy construction/casting *and* method forwarding for simple newtypes, this may work.

Consider this:

```
let MyInt257 = struct {
   inner: Int257

  impl Dereference(Int257) {
     fn dereference(self) { self.inner }
  }

  impl From(Int257) {
     fn from(i: Int257) { MyInt257{ inner: i } }
  }

  
}
```

and it'll allow `MyInt257` to be implicitly used as Int257, both as an underlying dereferenced type and to convert back from `Int257`, say, if we want to add a `MyInt257` to `Int257` and be it typed as `MyInt257` *or* `Int257` depending on expectations.

Please note that the exact syntax, naming and the API surface are just an approximation. I base this on an assumption that `impl` takes an expression that should resolve to an interface. For example, `From(Int257)` is simply this:

```
interface From(T: Type) {
  fn from(v: Int257) -> Self
}
```

or, without a shorthand,

```
let From = fn(T: Type) {
  interface {
   fn from(v: T) -> Self
  }
}
```

**Reply by p0lunin** [reply]
```
let MyInt257 = struct {
   inner: Int257

  impl Dereference(Int257) {
     fn dereference(self) { self.inner }
  }

  impl From(Int257) {
     fn from(i: Int257) { MyInt257{ inner: i } }
  }
}
```
There are few problems with such approach.

If we can declare fields whether we want - it will strongly decrease readibility because nothing keeps us from possibility to declare types like in follows snippet:
```
let MyInt257 = struct {
   inner: Int257

   impl Dereference(Int257) {
     fn dereference(self) { self.inner }
   }

   impl From(Int257) {
     fn from(i: Int257) { MyInt257{ inner: i } }
   }
   
   inner2: Int257
}
```
This is not a good way to declare data types because we never can be sure that we know all the information contained in the struct and it is hard to read and edit such types. So, there are few options to solve it:
1. Disallow to declare impls before struct fields. This is very bad way when we allow some syntatic construction that we disallow later - it is symptom of inconsistency in the language.
2. Allow to declare impls before struct fields but make warnings about this. But there are no guarantees from that user won't disable our warnings or that he will accept our warnings.

**Reply by cmxog** [reply]
We already do (1)

**Comment by p0lunin** [comment]
My proposal is to define `impl` as a sugar for the functions. 

## `@method` annotation
Core part of this approach consists of one annotation `@method`. This function will add current function to the namespace of the item should implement such method. To be more clear, see following fragment:
```
@method(Struct)
fn do_stuff(self: Struct, ...) { ... }
// generic version
@method(generics: [T, U, ...], Struct(T, U, ...)) // suppose that `[]` is an array or a list
fn do_stuff(self: Struct(T, U, ...), ...) { ... }
// generics are forbidden to use in such way
@method(generics: [T], T)
fn do_stuff(self: Struct(T, U, ...), ...) { ... }
```
In case of such syntax name lookup semantic when calling `value.method()` will be as follows:

1. `@method` annotation is a built-in method that add to the compiler information about that `do_stuff` method can be associated with the `Struct` data type if it satisfies generics conditions. You can say that this is same as constraint-based system as in Rust from which we are trying to move away, but we need constraints for generics because they are generics.
2. When compiler find `value.method()` call it tries to find information about associated with `type_of(value)` methods.
3. If compiler find methods then it will try to satisfy conditions that are declared in `@method` call (such as generics and requirements of generics implementations). If it cannot find such it will throw an compile error.

## `@impements` for interfaces
For interface impl this proposal provide addition annotation `@implements`:
```
@impements(Interface1, for: Struct)
fn do_stuff(self: Struct, ...) { ... }
// generic version
@impements(generics: [T], Interface1(T), for: Struct)
fn do_stuff(self: Struct(T), ...) { ... }
// free function version
@impements(Interface1, for: Struct)
fn do_stuff(...) { ... }
```
As you can see, user in this case can declare only part of interface methods so compiler cannot guarantee that user will implement all interface methods which can be a problem. There are few possible ways to fix this issue:
1. Compile-time checks that interface is implemented if it declared to be implemented.
2. Compile-time checks in the sugared version of this annotation (see next section).

## `impl` as a sugar
Since it can be annoying to write such annotations every time when you need to define a method, we can add syntatic sugar that looks like Rusts impl:
```
impl Struct { ... }
impl(T) Struct(T) { ... } // generic version
impl(T) Interface(T) for Struct(T) { ... } // generic version for interface impls
impl(T) Interface(T) for T { ... } // forbidden
```

**Comment by p0lunin** [comment]
After private discusses we have come to the following system:

Impls are consolidated with struct/union definition, where impl is defined with keyword `impl` as in follows:
```
struct T {
  /* fields and functions */
  impl Interface {
    /* functions */
  }
}
```

Consolidated syntax is required because of Tact computation semantics: types can be returned from functions, and it is confusing if impls located outside of struct definition: how then can be executed and attached to the struct definitions? 

EBNF of the impls:
```ebnf
impl = "impl", expr, "{", { sugared fn def }, "}";
```

Extensions to the structs and unions syntax:
```ebnf
(* Before *)
struct body = "{", { struct field }, { sugar fn def }, "}";
union body = "{", { union case }, { sugar fn def }, "}";
(* After *)
struct body = "{", { struct field }, { sugar fn def }, { impl }, "}";
union body = "{", { union case }, { sugar fn def }, { impl }, "}";
```
